### ISSUE TITLE
Resource Indicators for OAuth 2.0 (RFC 8707)

### ISSUE THREAD

**Issue by yohny** [keycloak/keycloak#14355]
### Description

Currently Keycloak supports `audience` parameter to indicate the target resource (client) that the token will be used for (https://www.keycloak.org/docs/latest/securing_apps/#form-parameters). However according to [RFC 8707](https://www.rfc-editor.org/rfc/rfc8707) the parameter to indicate target resource should be `resource`.

### Motivation

improve standard compliance

**Comment by sschu** [comment]
@yohny I think your wording is a bit misleading. Keycloak uses the audience claim to denote the resource server that should use it to check whether the token is intended for him. Having an additional optional resource parameter according to RFC 8708 would be an additional feature, the audience claim would still stay the same.

**Comment by yohny** [comment]
isn't that exactly what rfc 8707 says that `resource` parameter should be used for?
the resource server should always check the `aud` claim (inteneded audience) of a token to check whether the token is for it
the rfc 8707 specifies that client should use `resource` parameter when requesting tokens to idicate for which resource server (another client) the token will be used for...and as far as I can tell this is what `audience` parameter is doing now, or am I missing something?

**Comment by sschu** [comment]
Its actually giving the resource server (aud) an indication of which resource on the resource server.

**Comment by sschu** [comment]
Its actually giving the resource server (aud) an indication of which resource on the resource server - assuming there are multiple resources. It is somewhat similar to the scope parameter.

**Comment by yohny** [comment]
I might be reading it wrong, but I dont see anywhere in the specification that `resource` should indicate a  "sub-resource" within a resource server; quoting from spec:

> When the resource parameter is used in an authorization request to the authorization endpoint, it indicates the identity of the protected resource(s) to which access is being requested

> When the authorization server is informed of the resource that will process the access token, it can restrict the intended audience of that token to the given resource such that the token cannot be used successfully at other resources
 
and specific [section about access token](https://www.rfc-editor.org/rfc/rfc8707#section-2.2-1)
> When the resource parameter is used on an access token request made to the token endpoint, for all grant types, it indicates the target service or protected resource where the client intends to use the requested access token.

to me that seems quite similar to current `audience` behavior, with the distinction that you dont put there `client id`, but absolute URI instead (which can be then mapped to client id, as stated in the last sentence in the quote below)

> The authorization server SHOULD audience-restrict issued access tokens to the resource(s) indicated by the resource parameter. Audience restrictions can be communicated in JSON Web Tokens with the aud claim and the top-level member of the same name provides the audience restriction information in a Token Introspection response. **_The authorization server may use the exact resource value as the audience or it may map from that value to a more general URI or abstract identifier for the given resource._**

**Comment by sschu** [comment]
In my understanding, RFC 8707 is an addition to what is currently available. It won't replace what is currently there, especially not the audience claim. So if you have a good use case why you need this supported, just go ahead.

**Comment by yohny** [comment]
My intention is not to replace `audience` parameter as that would be a breaking change to many clients. Instead I would like the `resource` parameter to be added (perhaps as an alias to `audience`) to have better compliance with standards.

**Comment by sschu** [comment]
That's totally fine. You might want to start a discussion first to check if anybody else is interested and if you could join forces to provide an implementation for this.

**Comment by spaceone** [comment]
> first to check if anybody else is interested
I am interested!

**Comment by stianst** [comment]
Resource indicators would be useful to a client that is accessing multiple resource providers to be able to obtain tokens with a smaller scope basically. Very similar to `scope` just with a bit of a different syntax.

For example myclient could have access to api-1 and api-2, which would result in access tokens having `aud: [api-1, api-2]'. However, myclient doesn't always want to send an access token to `api-1` that `api-1` could then use to access `api-2`. 

In addition it can also scope a token to a sub-resource within the resource server.

I'd argue that `resource` should not be included in the access token, but rather used to control what `aud` and `scopes` are included in the token. Making it transparent to the resource providers.

One problem is that it leverages URIs instead of client-ids, which means there would need to be some mapping. That could be done through looking at the clients configured for the full audience, and removing any that doesn't have a URL that matches the URIs requested in the `resource` parameter.

To allow sub-resources I would imagine you'd want to map a sub-resource to a scope rather than an audience. So for example `resource=https://myapi` could result in an access token with aud=myapi, while `resource=https://myapi/galleries` would result in an access token with `aud=myapi` and scope `galleries`, or something along those lines.

It's also worth mentioning that this use-case can also be handled with scopes themselves in Keycloak. You can have scopes like `api-1`, `api-2`, `api-2:galleries`, and then rather than using `resources` use `scope`. Not sure if Keycloak supports this, but strictly speaking according to the spec you can use a refresh token to obtain an access token with a smaller scope.

End of the day I think we could support `resource` param as well in Keycloak fairly easily by just mapping it to `aud` and scopes.

**Comment by cgeorgilakis** [comment]
@stianst @mposolda as I understand Keycloak intends to support RFC 8707 Resource Indicators for OAuth 2.0 as a protocol mapper ( resource to aud and/or scope). This is great news. 
We have opened a PR with different logic for this. Now, we can not procceed with a PR with the logic you mentioned.
However, we are also interesting in this feature and it would be useful if someone could make a PR.

We want only to mention that we believe that aud could be either the resource as it is (URL) or a mapping client id as you mentioned.
This is based on the[ RFC 8707 specification](https://datatracker.ietf.org/doc/html/rfc8707#name-resource-parameter) : 
> The authorization server may use the exact resource value as the audience or it may map from that value to a more general URI or abstract identifier for the given resource.

Moreover, [JSON Web Token (JWT) Profile for OAuth 2.0 Access Tokens](https://datatracker.ietf.org/doc/html/rfc9068) mentions aud as required parameter in access token. We have proposed also Keycloak supporting this with an optional default value (no migration needed). This was part of our related closed issue. Do you want to open a new issue for this?

**Comment by rcarpa** [comment]
I'm reading this discussion and it confuses me a bit. It seems like keycloak correctly [implements](https://github.com/keycloak/keycloak/blob/90bf88c5400482d3b581c2dd8953a04a3f722800/services/src/main/java/org/keycloak/protocol/oidc/DefaultTokenExchangeProvider.java#L297) the `audience` parameter for token exchange, according to [RFC8693](https://datatracker.ietf.org/doc/html/rfc8693#name-request). This is a completely different problem from implementing [RFC8707](https://datatracker.ietf.org/doc/html/rfc8707#name-resource-parameter), which extends [RFC6749](https://datatracker.ietf.org/doc/html/rfc6749#section-3.2) to add the `resources` parameter to the authorization requests and token requests. Why did RFC8707 and RFC8693 decide on two different parameter names and format for basically the same goal is a mystery for me. However, just making `audience` a synonym to `resources` (as proposed in this discussion thread) doesn't seem correct.

Keycloak also [implements](https://github.com/keycloak/keycloak/blob/0e535d2bbec242c85a9ac00995d832d33087ef53/services/src/main/java/org/keycloak/protocol/oidc/endpoints/TokenEndpoint.java#L976) the `audience` field for the [UMA](https://docs.kantarainitiative.org/uma/wg/rec-oauth-uma-grant-2.0.html#rfc.section.3.3.1) grant type. This is probably a non-standart extension, because I couldn't find any standart mentioning it.  This is probably similary to how auth0 [allows](https://auth0.com/docs/secure/tokens/access-tokens/get-access-tokens#example-post-to-token-url) the audience parameter for the token requests without it being part of any standard I could find (if you know about one, please point it to me).

**Comment by thomasdarimont** [comment]
FYI I gave this another spin: https://github.com/keycloak/keycloak/pull/35711
The PR adds support for the resource parameter and allows clients to use a new `ResourceIndicatorMapper` OIDC mapper to limit the allowed resource indicators for a client. The `ResourceIndicatorMapper` also adds the requested allowed resources to the `aud` claim of the token.

Note that the `ResourceIndicatorMapper` supports URIs (following RFC8707) or Keycloak `client_id`s as allowed resource indicators. Allowed resource indicators can be explicit clientIds / URIs or regex patterns.

More details can be found in the description of the PR.

**Comment by stianst** [comment]
In my mind there are two levels of supporting resource indicators;

1. Resource Server
2. Resources within Resource Servers

For 1 we could leverage the fact that a client has a Root URL or perhaps introduce a separate Resource URL (https://www.ietf.org/archive/id/draft-ietf-oauth-resource-metadata-13.txt does provide a resource field for resource servers, this is something we could consider as well). We don't currently have a strict binding between clients and resource servers, but rather an indirect through either client roles or protocol mappers and audience claims. We should most likely add a proper way to configure what clients a client can connect to. In any case, essentially the logic would be something like:

Identity resource servers the client can connect to
For each resource parameter match them up against the Resource URL of those clients
That would be what I would prioritise supporting around resource indicators initially. At a later stage we could extend on this to include 2 from above; in which case we'd need some way of enumerating resources within a resource server; we could have an option to only check all resource parameters start with the Resource URL of a client; or to actually verify all resources exists. We could even go as far as leveraging authorization services to check that the user should have access to the specific resource or not. However, I would for sure start with something simple and only consider what I laid out for part 1.

**Comment by thomasdarimont** [comment]
I think using the root URL for clients representing resource servers as a "default" resource indicator is a good idea. However,  a dedicated attribute like a `Resource URI` might help us avoid some conflicts with existing code that derives other URLs from the root url if configured.

Regarding `Identity resource servers the client can connect to` we could introduce a  `OAuth2ResourceIndicatorResolver` SPI, which can be used to resolve the resource indicators a particular client has access to. 
To enumerate the resources available for a resource server users could either specify a `OAuth2ResourceIndicatorResolver` for the client, or use the Resource management of the Authorization Services which already allows to manage "resource server" client scoped resources.

**Comment by stianst** [comment]
> I think using the root URL for clients representing resource servers as a "default" resource indicator is a good idea. However,  a dedicated attribute like a `Resource URI` might help us avoid some conflicts with existing code that derives other URLs from the root url if configured.
> 
> Regarding `Identity resource servers the client can connect to` we could introduce a  `OAuth2ResourceIndicatorResolver` SPI, which can be used to resolve the resource indicators a particular client has access to. 
> To enumerate the resources available for a resource server users could either specify a `OAuth2ResourceIndicatorResolver` for the client, or use the Resource management of the Authorization Services which already allows to manage "resource server" client scoped resources.

Yeah, let's just go with a separate resource Uri. Maybe resource root Uri?! That matches the resource server metadata spec anyways 

I like the idea of an SPI for that for sure, and make it configurable on a per client basis.

I would suggest we break this up into two separate things; sort out 1 first which should be relatively simple, then address 2 as a follow-up.

**Comment by stianst** [comment]
This is related to something new we plan to add to Keycloak; have the ability to link clients to services. Once that is done this should be relatively simple to achieve as we would know what services a client can connect to and can check if the resource parameter matches the `Resource URI` of any of those services.

**Comment by cgeorgilakis** [comment]
From [JSON Web Token (JWT) Profile for OAuth 2.0 Access Tokens](https://datatracker.ietf.org/doc/html/rfc9068) : 

> aud  REQUIRED - as defined in [Section 4.1.3 of [RFC7519]](https://datatracker.ietf.org/doc/html/rfc7519#section-4.1.3).  See
      [Section 3](https://datatracker.ietf.org/doc/html/rfc9068#section-3) for indications on how an authorization server should
      determine the value of "aud" depending on the request.
> If the request does not include a "resource" parameter, the
   authorization server MUST use a default resource indicator in the
   "aud" claim.  If a "scope" parameter is present in the request, the
   authorization server SHOULD use it to infer the value of the default
   resource indicator to be used in the "aud" claim. 

Based on that aud default value is needed if not requested using resource indicators. Default aud could be default empty if someone does not want it and for keep Keycloak behaviour same as old versions. Current audience mapper is not enough.

We propose to be realm setting or extend audience mapper for adding only default aud value not in every case.

**Comment by stianst** [comment]
Description updated, including comparing resource indicators to existing mechanisms in Keycloak for audience scoping tokens.

**Comment by stianst** [comment]
Alternative proposal for resource indicator support aiming to be more aligned with existing mechanisms for audience scoping tokens in Keycloak:
https://github.com/keycloak/keycloak/discussions/35743#discussioncomment-15980055
